src/serac/physics/heat_transfer.hpp

```diff
@@ -237,34 +262,28 @@ class HeatTransfer<order, dim, Parameters<parameter_space...>, std::integer_sequ
   template <int... active_parameters, typename MaterialType>
   void setMaterial(DependsOn<active_parameters...>, MaterialType material)
   {
-    K_functional_->AddDomainIntegral(
-        Dimension<dim>{}, DependsOn<0, active_parameters + 1 ...>{},
-        [material](auto x, auto temperature, auto... params) {
+    residual_->AddDomainIntegral(
+        Dimension<dim>{}, DependsOn<0, 1, 2, active_parameters + NUM_STATE_VARS...>{},
+        [material](auto x, auto temperature, auto dtemp_dt, auto shape, auto... params) {
           // Get the value and the gradient from the input tuple
-          auto [u, du_dx] = temperature;
-          auto source     = serac::zero{};
-
-          auto response = material(x, u, du_dx, params...);
+          auto [u, du_dX] = temperature;
+          auto [p, dp_dX] = shape;
+          auto du_dt      = get<VALUE>(dtemp_dt);
 
-          return serac::tuple{source, -1.0 * response.heat_flux};
-        },
-        mesh_);
-
-    M_functional_->AddDomainIntegral(
-        Dimension<dim>{}, DependsOn<0, active_parameters + 1 ...>{},
-        [material](auto x, auto d_temperature_dt, auto... params) {
-          auto [u, du_dx] = d_temperature_dt;
-          auto flux       = serac::zero{};
-
-          auto temp      = u * 0.0;
-          auto temp_grad = du_dx * 0.0;
+          // Note that the current configuration x = X + p, where X is the original reference
+          // configuration and p is the shape displacement. We need the gradient with
+          // respect to the perturbed reference configuration x = X + p for the material model. Therefore, we calculate
+          // du/dx = du/dX * dX/dx = du/dX * (dx/dX)^-1 = du/dX * (I + dp/dX)^-1
 
-          auto response = material(x, temp, temp_grad, params...);
+          auto du_dx = dot(du_dX, inv(I + dp_dX));
 
-          auto source = response.specific_heat_capacity * response.density * u;
+          auto response = material(x + p, u, du_dx, params...);
+          auto source   = response.specific_heat_capacity * response.density * du_dt;
 
-          // Return the source and the flux as a tuple
-          return serac::tuple{source, flux};
+          // Note that the return is integrated in the perturbed reference
+          // configuration, hence the det(I + dp_dx) = det(dx/dX)
+          return serac::tuple{source * det(I + dp_dX),
+                              -1.0 * dot(response.heat_flux, inv(transpose(I + dp_dX))) * det(I + dp_dX)};
```

Comment:
I think we can simplify the long `dot(...)` expression slightly from

```
dot(response.heat_flux, inv(transpose(I + dp_dX)))
```
to
```
dot(inv(I + dp_dX), response.heat_flux)
```
-----------------

also, `det(I + dp_dX)` and `inv(I + dp_dX)` appear a couple of times in this kernel. It's probably worth only computing them once and storing it as temporary. I'm not sure what a good name for that matrix `I + dp_dX` would be.